src/globalvars.cpp

```diff
@@ -77,17 +134,130 @@ void update_global_vars(const struct bpf_object *bpf_object,
 
   // Update the values for the global vars (using the above offsets)
   for (auto [name, offset] : vars_and_offsets) {
-    if (offset < 0) {
-      LOG(BUG) << "Global variable has not been added to the BPF code "
-                  "(codegen_llvm)";
-    }
-
     int64_t *var = reinterpret_cast<int64_t *>(global_vars_buf + offset);
 
-    if (name == NUM_CPUS) {
+    auto global_var = from_string(name);
+    if (global_var == GlobalVar::NUM_CPUS) {
       *var = bpftrace.ncpus_;
+    } else if (global_var == GlobalVar::MAX_CPU_ID) {
+      *var = bpftrace.max_cpu_id_;
     }
   }
 }
 
+static void update_global_vars_custom_rw_section(
+    const struct bpf_object *bpf_object,
+    const std::string &section_name,
+    struct bpf_map *global_vars_map,
+    const std::unordered_set<std::string> &needed_global_vars,
+    BPFtrace &bpftrace)
+{
+  if (needed_global_vars.size() > 1) {
+    LOG(BUG) << "Multiple read-write global variables are in same section "
+             << section_name;
+  }
+  auto name = *needed_global_vars.begin();
+  auto config = get_config(name, bpftrace.resources);
+  auto desired_size = (bpftrace.max_cpu_id_ + 1) * config.type.GetSize();
```

Comment:
Just throwing this out there: from your example if there were 64 cpus but only one is available/online #63 - then we're allocating 63 additional slots for cpus that we'll never hit. Maybe we should only allocated based on num cpus and the create a way to map the CPU id to an open slot index.

Reply:
> Just throwing this out there: from your example if there were 64 cpus but only one is available/online https://github.com/bpftrace/bpftrace/issues/63 - then we're allocating 63 additional slots for cpus that we'll never hit.

Agreed, that is annoying.

> Maybe we should only allocate based on num cpus and the create a way to map the CPU id to an open slot index.

Ah I see, we could allocate a global uint8 array of size N = max_cpu_id + 1, then in each index, we have the open slot index. So in that example, we map 0-62 -> -1 (invalid) and 63 -> 0.

That could work! It certainly avoids the memory over-allocation but involves another array lookup and likely a CreateSelect to bound the result.

Do you know if per-CPU arrays in BPF already allocate memory for each possible CPU or only the online ones? If the former, then this is equivalent for what we have now and I would prefer to do it in another PR. According to this comment thread, it is the former: https://github.com/iovisor/bcc/issues/893. Happy to discuss of course!